### PR TITLE
Project: Associations

### PR DESCRIPTION
**Pull request by verissimohenry**
In this project, I build an application named Private Events and that behaves similarly to Eventbrite. The main goal was to put in practice the main concepts of Associations in rails.

### CODE REVIEW

**Review by NtwaliHeritier — CHANGES_REQUESTED** [review]
### **STATUS: INVALID❌❌❌❌**

Hello @verissimohenry 
Great work submitting this project.
I can tell you put in a lot of work, however, It was invalidated due to the following reason.

When I run the project locally, I get the error below. Kindly fix it and ask for another review👍

Capture

Happy coding💻💻💻

**Review by george-shammar — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUIRED** :recycle: :recycle: 

This is impressive. :+1: :clap: 

**Highlights:**
- Your models and associations are set up correctly. :+1: 
- Credit for extra step you took to make your user interface neat.

However, you need to make some changes in order to get the project approved.

- [x] The event's show page should display the list of attendants. I created an event and invited two users but the event show page is not displaying any of them.
- [x] User should be able to retrieve future events the user has created as well as past events the user has created. Currently, both are lumped together in the user's show page.
- [x] Your project has no README. Please include one.
- [x] Include validation and association tests for your models. You currently have none.
- [x] Authentication and event management features too should have integration tests. Make use of Capybara.
- [x] Remove commented codes from your repo. It is not professional. e.g Invites_controller.rb and everywhere else you have commented codes.

[OPTIONAL]

- [x] Specify routes only for used actions. e.g You have ` resources :invites` whereas you did not make use of all the controller actions.

Effect these changes and submit another CRR when you're ready.

If you require clarification, drop your message here and ping me on slack.

Cheers and happy coding. :+1: :smile:

**Review by taiwo2 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️
Hey @verissimohenry, Weldone for implementing this changes:
#### Highlights:
- [x] The event's show page should display the list of attendants. I created an event and invited two users but the event show page is not displaying any of them.
- [x]  User should be able to retrieve future events the user has created as well as past events the user has created. 

However this changes are still required to get youer project approved:
#### Improve:
- [x] Your project has no README. Please include one.
- [x]  Include validation and association tests for your models. You currently have none.
- [x]  Authentication and event management features too should have integration tests.
- [x]  Remove commented codes from your repo. 
Once Done kindly requested for another review

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, thank you for working on changes suggested by previous reviewers :ok_hand:
## To Highlight 🎉
- [x] You have mastered Git Flow.
- [x] PR is descriptive.
## To Improve 🙏
- [x] Please add a README file. KIndly see this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_readme.png) and this [lesson](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/) from Pathwright.
- [x] Kindly remove commented code from invites controller. It makes code unclean.
### User's show action:
- [x] It should list events created by user, i.e. @created_events = current_user.events. Please see this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png).
 - [x] It should display events that the user is attending to separated in the past and upcoming.
- [x] It should list events previously attended by user using event model scope, i.e. @past_events = current_user.attended_events.previous. Please see this screenshot.
- [x] It should list upcoming events previously attended by user using event model scope, i.e. @upcoming_events = current_user.attended_events.upcoming. Please see this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png).
- [x] The event's show page should display the list of attendees.
- [x] I am unable to invite a user to attend an event I have created.
- [x] Please add unit tests for validations and associations of models using Rspec.
- [x] Please add integration tests for authentication using Rspec + Capybara.
- [x] Please add integration tests for event management features using Rspec + Capybara.
- [x]  Please make sure that you have worked on all the changes suggested and your project is ready for a review before asking for another code review.

# Optional
- [x] Please protect routes by configuring only those routes which are being used  i.e. resources :events, only: [:new, :create, :show, :index]

### Happy Coding ☺️

## If you have any questions you can comment here in this PR or can always connect with me on slack ( @Sajjad Ahmad ).

**Comment by mke2111** [comment]
## Project is under review. Please be patient

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done lad and thanks for implementing the changes requested earlier on**. 

### Well-done 💯 
    1. The app runs just fine and a user can be invited successfully.
    2. The readme file is now clear and concise

**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

### REQUIRED

- [x] I'm unable to log in with an existing username. A pop-up of **Invalid name** is shown. Give this a fix.
- [x] You currently have 3 unit tests. Add more unit(validation and association) and integration(feature, requests) tests. You may choose to use Shoulda, CapyBara, Rspec gems.
- [x] You have numerous pieces of commented-out code in your files. (invites controller) Clean up your repo.


#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

File: app/controllers/invites_controller.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  #   @event = Event.find(params[:event_id])
+  #   @invite = Invite.find(@event.id)
+  #   @invite.update(discription: true)
+  #   redirect_to root_path
+  # end
+
+  # def deactivate
+  #   @event = Event.find(params[:event_id])
+  #   @invite = Invite.find(@event.id)
+  #   @invite.update(discription: false)
```
Comment (mke2111) [inline app/controllers/invites_controller.rb]: Clean up this file

File: README.md [hunk 2]
```diff
@@ -5,7 +5,112 @@ application up and running.
...
+## Rspec testing
+ > To test all models and scopes, write the following in your terminal:
+ ```
+$ rspec spec/models/user_spec.rb
+```
+```
+$ rspec spec/models/event_spec.rb
+```
+```
+$ rspec spec/models/invite_spec.rb
```
Comment (mke2111) [inline README.md]: Add tests in all these files.

File: README.md [hunk 3]
```diff
@@ -5,7 +5,112 @@ application up and running.
...
+- Event show page by clicking on the show button on your event card.
+- Choose friend's name from the select options and click invite.
+
+## Rspec testing
+ > To test all models and scopes, write the following in your terminal:
+ ```
+$ rspec spec/models/user_spec.rb
+```
+```
+$ rspec spec/models/event_spec.rb
```
Comment (mke2111) [inline README.md]: OPTIONAL

- [x] You may choose to run `rspec` to run all the test files in the spec directory

File: app/controllers/invites_controller.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      render :new
+    end
+  end
+
+  # def activate
+  #   @event = Event.find(params[:event_id])
+  #   @invite = Invite.find(@event.id)
+  #   @invite.update(discription: true)
+  #   redirect_to root_path
+  # end
```
Comment (rindrajosia) [inline app/controllers/invites_controller.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:
To Highlight: :angel:

- It has functional login page :+1: 
- Logged in user can create new event :100: 
- Navigation links are present at the top of the page :1st_place_medal: 
- Event's index page displays past and upcoming events lists :+1: 
- The event's show page displays the list of attendants :100: 
- The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming :1st_place_medal: 

To improve :factory_worker: [REQUIRED]:

- [x] Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.
- [x] models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
- [x] authentication have has integration tests (Rspec + Capybara) ( [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
- [x]  event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

Nice to have:

- [ ] routes only for used actions, i.e. resources :events, only: [:new, :create, :show, :index]

That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

**Review by umairarshadbutt — APPROVED** [review]
# **Status: PROJECT APPROVED** :champagne:
Excellent job on this project guys. :fireworks: Keep it up.
##
**Hi @verissimohenry  It is the 7th review, but there are still issues with testing. I will go ahead and mark this as complete.**
- You can now merge to master after completing the unit and integration testing and submit the project completion form :rocket: . I wish you all the very best in your next tasks :1st_place_medal:
##
HAPPY CODING :computer:
**[Umair Arshad (Code Reviewer)](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Pull request merged**